extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoDiscoveryParticipant.java

```diff
@@ -0,0 +1,115 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.binding.wemo.discovery;
+
+import static org.eclipse.smarthome.binding.wemo.WemoBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.binding.wemo.WemoBindingConstants;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.UpnpDiscoveryParticipant;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.jupnp.model.meta.RemoteDevice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link WemoDiscoveryParticipant} is responsible for discovering new and
+ * removed Wemo devices. It uses the central {@link UpnpDiscoveryService}.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ * @author Kai Kreuzer - some refactoring for performance and simplification
+ *
+ */
+public class WemoDiscoveryParticipant implements UpnpDiscoveryParticipant {
+
+    private Logger logger = LoggerFactory.getLogger(WemoDiscoveryParticipant.class);
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return WemoBindingConstants.SUPPORTED_THING_TYPES;
+    }
+
+    @Override
+    public DiscoveryResult createResult(RemoteDevice device) {
+        ThingUID uid = getThingUID(device);
+        if (uid != null) {
+            Map<String, Object> properties = new HashMap<>(2);
+            String label = "WeMo Device";
+            try {
+                label = device.getDetails().getFriendlyName();
+            } catch (Exception e) {
+                // ignore and use default label
+            }
+            properties.put(UDN, device.getIdentity().getUdn().getIdentifierString());
+
+            DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(label)
+                    .withRepresentationProperty(UDN).build();
+
+            logger.debug("Created a DiscoveryResult for device '{}' with UDN '{}'",
+                    device.getDetails().getFriendlyName(), device.getIdentity().getUdn().getIdentifierString());
+
+            return result;
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public ThingUID getThingUID(RemoteDevice device) {
+        if (device != null) {
+            if (device.getDetails().getManufacturerDetails().getManufacturer() != null) {
+                if (device.getDetails().getManufacturerDetails().getManufacturer().toUpperCase().contains("BELKIN")) {
+                    if (device.getDetails().getModelDetails().getModelName().toLowerCase().startsWith("socket")) {
```

Comment:
I have seen UPnP devices where `getDetails().getModelDetails().getModelName()` leads to an NPE, because they lack the information - it is therefore better to do rigid null checks on such paths (or catch NPEs around the whole code block)